In [42]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
%matplotlib inline 

In [43]:
grade_df = pd.read_csv('student-mat.csv')

In [44]:
grade_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null int64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null int64
Fedu          395 non-null int64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null int64
studytime     395 non-null int64
failures      395 non-null int64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null int64
freetime      395 non-null int64
goout         395 non-null int64
Dalc          395 no

In [45]:
grade_df = pd.concat([grade_df,pd.get_dummies(grade_df.sex, prefix="sex", drop_first=True)], axis=1)

In [46]:
grade_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 34 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null int64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null int64
Fedu          395 non-null int64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null int64
studytime     395 non-null int64
failures      395 non-null int64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null int64
freetime      395 non-null int64
goout         395 non-null int64
Dalc          395 no

In [47]:
#target variable is Final Grade

In [48]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt

In [49]:
grade_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 34 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null int64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null int64
Fedu          395 non-null int64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null int64
studytime     395 non-null int64
failures      395 non-null int64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null int64
freetime      395 non-null int64
goout         395 non-null int64
Dalc          395 no

In [50]:
#OLS 
X = grade_df[['absences', 'G1','G2']]
Y = grade_df.G3

In [51]:
import statsmodels.api as sm
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

//anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     G3   R-squared:                       0.826
Model:                            OLS   Adj. R-squared:                  0.825
Method:                 Least Squares   F-statistic:                     619.5
Date:                Fri, 14 Feb 2020   Prob (F-statistic):          3.90e-148
Time:                        21:31:15   Log-Likelihood:                -815.59
No. Observations:                 395   AIC:                             1639.
Df Residuals:                     391   BIC:                             1655.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.0675      0.341     -6.060      0.000      -2.738      -1.397
absences       0.0364      0.012      3.010      0.003       0.013       0.060
G1             0.1545      0.056      2.779      0.006       0.045       0.264
G2             0.9884      0.049     20.142      0.000       0.892       1.085
==============================================================================
Omnibus:                      228.806   Durbin-Watson:                   1.827
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1474.871
Skew:                          -2.493   Prob(JB):                         0.00
Kurtosis:                      11.046   Cond. No.                         60.9
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [52]:
#KNN

In [53]:
X = pd.concat([grade_df.G1,grade_df.G2,grade_df.absences], axis=1)

In [54]:
len(pd.DataFrame([grade_df.G1,grade_df.G2,grade_df.absences]))


3

In [55]:
len(grade_df.G3)

395

In [56]:
X

G1  G2  absences
0     5   6         6
1     5   5         4
2     7   8        10
3    15  14         2
4     6  10         4
5    15  15        10
6    12  12         0
7     6   5         6
8    16  18         0
9    14  15         0
10   10   8         0
11   10  12         4
12   14  14         2
13   10  10         2
14   14  16         0
15   14  14         4
16   13  14         6
17    8  10         4
18    6   5        16
19    8  10         4
20   13  14         0
21   12  15         0
22   15  15         2
23   13  13         0
24   10   9         2
25    6   9        14
26   12  12         2
27   15  16         4
28   11  11         4
29   10  12        16
..   ..  ..       ...
365  10  10         4
366  13  13         0
367   7   6         0
368  11  10         0
369  14  12        10
370   7   7         4
371  14  12         3
372  13  11         8
373   6   5        14
374  19  18         0
375   8   8         2
376  15  14         4
377   8   9         4
378  15  15         0
379  10  10        17
380  15  14         4
381   7   6         5
382  11  11         2
383   6   5         0
384   6   5        14
385  10   9         2
386   6   5         7
387   7   5         0
388   7   9         0
389   6   5         0
390   9   9        11
391  14  16         3
392  10   8         3
393  11  12         0
394   8   9         5

[395 rows x 3 columns]

In [57]:
from sklearn import neighbors

# Build our model.
knn = neighbors.KNeighborsRegressor(n_neighbors=10)
X = pd.concat([grade_df.G1,grade_df.G2,grade_df.absences], axis=1)
Y = grade_df.G3
knn.fit(X, Y)



KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
                    metric_params=None, n_jobs=None, n_neighbors=10, p=2,
                    weights='uniform')

In [58]:
T = np.arange(2, 30, 0.1)[:, np.newaxis]

In [61]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2)


In [64]:
knn = neighbors.KNeighborsRegressor(n_neighbors=10)
X = pd.concat([grade_df.G1,grade_df.G2,grade_df.absences], axis=1)
Y = grade_df.G3
knn.fit(X, Y)

KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
                    metric_params=None, n_jobs=None, n_neighbors=10, p=2,
                    weights='uniform')

In [67]:
from sklearn.model_selection import cross_val_score
score = cross_val_score(knn, X, Y, cv=10)
print("Unweighted Accuracy: %0.2f (+/- %0.2f)" % (score.mean(), score.std() * 2))
score_w = cross_val_score(knn, X, Y, cv=10)
print("Weighted Accuracy: %0.2f (+/- %0.2f)" % (score_w.mean(), score_w.std() * 2))

Unweighted Accuracy: 0.82 (+/- 0.17)
Weighted Accuracy: 0.82 (+/- 0.17)


The models are very similar. The OLS shows r-squared values at 82% and the accuracy rate for the KNN model is at 82% also. I think that in this instance the OLS was a better choice because my variables were based on the theory that together they assume a linear relationship with Y(G3). Therefore, I think that producing the OLS was more telling in helping understand the impact of each feature. I think that the KNN modeling would work best in situations where there aren't assumptions about a direct relationship like if we were forecasting weather as non-meterologists.